mlir/include/mlir/Dialect/MQTOpt/Transforms/Transpilation/Router.h

```diff
@@ -13,63 +13,38 @@
 #include "mlir/Dialect/MQTOpt/Transforms/Transpilation/Architecture.h"
 #include "mlir/Dialect/MQTOpt/Transforms/Transpilation/Common.h"
 #include "mlir/Dialect/MQTOpt/Transforms/Transpilation/Layout.h"
-#include "mlir/Dialect/MQTOpt/Transforms/Transpilation/Scheduler.h"
 
 #include <algorithm>
+#include <llvm/ADT/DenseSet.h>
 #include <mlir/Support/LLVM.h>
+#include <optional>
 #include <queue>
-#include <stdexcept>
 #include <utility>
 #include <vector>
 
 namespace mqt::ir::opt {
 
-/**
- * @brief A vector of SWAPs.
- */
-using RouterResult = SmallVector<QubitIndexPair>;
-
-/**
- * @brief A planner determines the sequence of swaps required to route an array
-of gates.
-*/
-struct RouterBase {
-  virtual ~RouterBase() = default;
-  [[nodiscard]] virtual RouterResult route(const Layers&, const ThinLayout&,
-                                           const Architecture&) const = 0;
-};
-
-/**
- * @brief Use shortest path swapping to make one gate executable.
- */
-struct NaiveRouter final : RouterBase {
-  [[nodiscard]] RouterResult route(const Layers& layers,
-                                   const ThinLayout& layout,
-                                   const Architecture& arch) const override {
-    if (layers.size() != 1 || layers.front().size() != 1) {
-      throw std::invalid_argument(
-          "NaiveRouter expects exactly one layer with one gate");
-    }
-
-    /// This assumes an avg. of 16 SWAPs per gate.
-    SmallVector<QubitIndexPair, 16> swaps;
-    for (const auto [prog0, prog1] : layers.front()) {
-      const auto [hw0, hw1] = layout.getHardwareIndices(prog0, prog1);
-      const auto path = arch.shortestPathBetween(hw0, hw1);
-      for (std::size_t i = 0; i < path.size() - 2; ++i) {
-        swaps.emplace_back(path[i], path[i + 1]);
-      }
+class NaiveRouter {
+public:
+  [[nodiscard]] static mlir::SmallVector<QubitIndexPair, 64>
+  route(QubitIndexPair gate, const ThinLayout& layout,
+        const Architecture& arch) {
+    mlir::SmallVector<QubitIndexPair, 64> swaps;
+    const auto hw0 = layout.getHardwareIndex(gate.first);
+    const auto hw1 = layout.getHardwareIndex(gate.second);
+    const auto path = arch.shortestPathBetween(hw0, hw1);
+    assert(path.size() >= 2 && "NaiveRouter::route: invalid shortest path");
+    for (std::size_t i = 0; i < path.size() - 2; ++i) {
+      swaps.emplace_back(path[i], path[i + 1]);
     }
```

Comment:
Just an optimization: Can we avoid creating two vectors here?
The current code generates the path as a vector only for it to group it up into pairs afterwards.
Certainly not a bottleneck, but feels like we could be doing better here.

Reply:
It could make sense to have an additional utility function in the architecture class that would directly return pairs.